### PR TITLE
Decompogen checking whether function contains `symbol` or not.

### PR DESCRIPTION
**Pull request by Shekharrajak**
Previous 

```
>>> print decompogen(sin(x)**2 + sin(x)-sqrt(3)/2 ,x)
[sin(x)**2 + sin(x) - sqrt(3)/2]
```

Now 

```
decompogen(sin(x)**2 + sin(x)-sqrt(3)/2 ,x) 
 [x**2 + x - sqrt(3)/2, sin(x)]
```

### CODE REVIEW

**Comment by Shekharrajak** [comment]
Ping @aktech

File: sympy/solvers/decompogen.py [hunk 1]
```diff
@@ -42,7 +43,7 @@ def decompogen(f, symbol):
 
     # ===== Convert to Polynomial ===== #
     fp = Poly(f)
-    gens = fp.gens
+    gens = list(filter(lambda x: symbol in x.atoms(Symbol) , fp.gens))
```
Comment (aktech) [inline sympy/solvers/decompogen.py]: `filter` already returns a tuple here, you don't need to typecast it into a list.

File: sympy/solvers/solvers.py [hunk 2]
```diff
@@ -1587,6 +1587,7 @@ def _expand(p):
                 u = False
             if u:
                 eq, cov = u
+                eq = simplify(eq)
```
Comment (aktech) [inline sympy/solvers/solvers.py]: Why this change?

File: sympy/solvers/tests/test_decompogen.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 from sympy.utilities.pytest import XFAIL
 
 x = Symbol('x')
+y = Symbol('y')
```
Comment (aktech) [inline sympy/solvers/tests/test_decompogen.py]: You can use `symbols` to wrap these two lines into one.

File: sympy/solvers/tests/test_decompogen.py [hunk 4]
```diff
@@ -12,6 +13,8 @@ def test_decompogen():
     assert decompogen(sqrt(6*x**2 - 5), x) == [sqrt(x), 6*x**2 - 5]
     assert decompogen(sin(sqrt(cos(x**2 + 1))), x) == [sin(x), sqrt(x), cos(x), x**2 + 1]
     assert decompogen(Abs(cos(x)**2 + 3*cos(x) - 4), x) == [Abs(x), x**2 + 3*x - 4, cos(x)]
+    assert decompogen(sin(x)**2 + sin(x)-sqrt(3)/2 ,x) == [x**2 + x - sqrt(3)/2, sin(x)]
```
Comment (aktech) [inline sympy/solvers/tests/test_decompogen.py]: Add space around minus operator in `sin(x)-sqrt(3)` .

Comment (aktech) [inline sympy/solvers/tests/test_decompogen.py]: `sqrt(3)/2 ,x`
Remove space after `sqrt(3)/2` and add space after comma.

**Comment by aktech** [comment]
Make the changes I suggested and squash all the commits into a single commit with a proper commit message. Then it's ready to go.

File: sympy/solvers/decompogen.py (hunk 1 above)
Comment (Shekharrajak) [inline sympy/solvers/decompogen.py]: Travis check failed ,then I added list there.
```
 if len(gens) == 1 and gens[0] != symbol:
    TypeError: object of type 'filter' has no len()
```
build was : https://travis-ci.org/sympy/sympy/jobs/110130814

Comment (aktech) [inline sympy/solvers/decompogen.py]: Oh! `filter` will return tuple only in Python 2 otherwise it will return an iterator. +1

Comment (smichr) [inline sympy/solvers/decompogen.py]: You should almost always be using something like `symbol in x.free_symbols` because `symbol` might appear in `x` as a dummy variable, e.g. `Integral(x, (x, 1, 2))` is independent of `x` -- it's not an x-dependent generator (if it appeared as a generator).

Comment (aktech) [inline sympy/solvers/decompogen.py]: +1

File: sympy/solvers/decompogen.py [hunk 5]
```diff
@@ -30,6 +30,7 @@ def decompogen(f, symbol):
     [x**2 - x - 1, x**2 + x]
 
     """
+    from sympy.core.symbol import Symbol
```
Comment (aktech) [inline sympy/solvers/decompogen.py]: This is not required now.

File: sympy/solvers/decompogen.py (hunk 1 above)
Comment (Shekharrajak) [inline sympy/solvers/decompogen.py]: Thanks @smichr

File: sympy/solvers/tests/test_decompogen.py [hunk 6]
```diff
@@ -12,7 +11,8 @@ def test_decompogen():
     assert decompogen(sqrt(6*x**2 - 5), x) == [sqrt(x), 6*x**2 - 5]
     assert decompogen(sin(sqrt(cos(x**2 + 1))), x) == [sin(x), sqrt(x), cos(x), x**2 + 1]
     assert decompogen(Abs(cos(x)**2 + 3*cos(x) - 4), x) == [Abs(x), x**2 + 3*x - 4, cos(x)]
-
+    assert decompogen(sin(x)**2 + sin(x) - sqrt(3)/2 ,x) == [x**2 + x - sqrt(3)/2, sin(x)]
```
Comment (aktech) [inline sympy/solvers/tests/test_decompogen.py]: See https://github.com/sympy/sympy/pull/10622#discussion_r53555965

File: sympy/solvers/tests/test_decompogen.py [hunk 7]
```diff
@@ -12,7 +11,8 @@ def test_decompogen():
...
+    assert decompogen(Abs(cos(y)**2 + 3*cos(x) - 4) ,x) == [Abs(x), 3*x + cos(y)**2 - 4, cos(x)]
```
Comment (aktech) [inline sympy/solvers/tests/test_decompogen.py]: See https://github.com/sympy/sympy/pull/10622#discussion_r53555965

Comment (Shekharrajak) [inline sympy/solvers/tests/test_decompogen.py]: Done.

**Comment by aktech** [comment]
Thanks for working on this. Looks good, Merging.
